### DISCUSSION TITLE
Refine with PostgREST

### DISCUSSION THREAD

**Discussion by gschader** [refinedev/refine#1394] [Q&A] [answered]
Hi. Started to work with refine and tried to use a REST API provided with PostgREST, which serves  RESTful API from any Postgres database. Is it correct that refine expects Id-attributes in every used resource? My API doesn't provide Id-attributes as the underlying tables don't provide them. Instead you can query the resources for example with the primary key attributes of the underlying tables.
As far as I've seen Refine doesn't support such a behavior, or? It seems possible to adjust the data-provider, but without Id-attribute most of the componenets, hooks etc. are not useable, I fear. Is this correct? Are their any solutions?
Thanks Günter

**Comment by omeraplak** [comment, marked as answer]
Hey @gschader ,
You can handle this with the `metaData` property in data hooks, forms, and tables.

For example,
if you don't have an id field in a table and you can pass your primary key attributes to `dataProvider` using `metaData`.

You can check the PostgREST dataProvider of `react-admin` for ideas,
https://github.com/raphiniert-com/ra-data-postgrest/blob/2c9ff6840122e059cc05910f61f1336b149d9225/src/dataProvider/index.ts

Please do not hesitate to contact us wherever you are stuck.

**Reply by gschader** [reply]
Hi @omeraplak ,
thanks for your proposal. I had already installed react-admin and also raphiniert for testing purposes before. Therefore I thought your proposal could be a good idea. Unfortunately now I have the following problem: I don't find an example how the metaData-property is build correctly. Is there any example code, where I can find more details on this type.?
Thanks Günter

**Reply by gschader** [reply]
Hi @omeraplak,
I fear that the implementation of a universal solution to use a postgREST-API with Refine means a lot of additional effort to customize Refine - and that is what I basically wanted to avoid when using a react-base framework like Refine.
One example: the Show-button is reposible to route from the list view to the single entity view, while within the single entity view (component) the useShow-hook is used, which identifies the entity id from the url. But for my postgREST-API much more than id is necessary to query a single entity. Ok, useShow is able to use metaData, but first I have to bring the meta content to the component. As Refine relies on "id" in the resources the complete process of data transfer seems much less complex.
Or do you have an idea to solve this with less effort?
Thanks Günter

**Reply by omeraplak** [reply]
Hey @gschader ,
I'm sorry for the delay.
If you can write a simple PostgREST data provider for Refine and let us know what is missing, let's see what we can do for you. Adding support for PostgREST isn't a feature we're planning anytime soon. But if Community wants to do it, we are happy to help. Please do not hesitate to contact us.

**Reply by fspijkerman** [reply]
@gschader PR #1490 introduces the capability to override the `select` attribute to the database (using the data provider Supabase)

https://postgrest.org/en/stable/api.html?highlight=join#embedding-through-join-tables

Example usage resource embedding using metaData select: 
```
const { tableProps, tableQueryResult } = useTable<IEvent>({
  resource: "events",
  metaData: {
    select: '*,devices(name)',
  },
});
```
